BackendServer/python/analysis.py

```diff
@@ -0,0 +1,163 @@
+import pandas as pd
+import numpy as np
+import mysql.connector
+from enum import Enum
+
+class AutoStation(Enum):
+    noPoints = 0
+    failed = 1
+    docked = 2
+    engaged = 3
+
+
+class TeleStation(Enum):
+    noPoints = 0
+    parked = 1
+    failed = 2
+    docked = 3
+    engaged = 4
+
+AUTO_ATTEMPTED_DOCK_STATES = (AutoStation.failed, AutoStation.docked, AutoStation.engaged)
+AUTO_SUCCESSFUL_DOCK_STATES = (AutoStation.docked, AutoStation.engaged)
+
+TELE_ATTEMPTED_DOCK_STATES = (TeleStation.failed, TeleStation.docked, TeleStation.engaged)
+TELE_SUCCESSFUL_DOCK_STATES = (TeleStation.docked, TeleStation.engaged)
+    
+# Safe division function to return None if division by zero so that it becomes NULL in the database
+def div(a, b):
+    if b == 0:
+        return None
+    return a / b
+
+# Does processing on a df of matches to get combined statistics
+def calculate_match_scores(matches_df, *, mutate=False):
+    if mutate:
+        output_df = matches_df
+    else:
+        output_df = matches_df.copy()
+
+    # Calculate cycle counts
+    output_df['Tele_Low_Cycles'] = output_df['Tele_Cube_Low'] + output_df['Tele_Cone_Low']
+    output_df['Tele_Mid_Cycles'] = output_df['Tele_Cube_Mid'] + output_df['Tele_Cone_Mid']
+    output_df['Tele_High_Cycles'] = output_df['Tele_Cube_High'] + output_df['Tele_Cone_High']
+    output_df['Tele_Total_Cycles'] =  output_df['Tele_Low_Cycles'] + output_df['Tele_Mid_Cycles'] + output_df['Tele_High_Cycles']
+
+    # Calculate grid points
+    output_df['Auto_Grid_Points'] = 3 * (output_df['Auto_Cube_Low'] + output_df['Auto_Cone_Low']) + 4 * (output_df['Auto_Cube_Mid'] + output_df['Auto_Cone_Mid']) + 6 * (output_df['Auto_Cube_High'] + output_df['Auto_Cone_High'])
+    output_df['Tele_Grid_Points'] =  2 * output_df['Tele_Low_Cycles'] + 3 * output_df['Tele_Mid_Cycles'] + 5 * output_df['Tele_High_Cycles']
+    
+    # Link points
+    # output_df['Assumed_Link_Points'] = 5 / 3 * (output_df['Auto_Cube_Low'] + output_df['Auto_Cone_Low'] + output_df['Tele_Cube_Low'] + output_df['Tele_Cone_Low'] + output_df['Auto_Cube_Mid'] + output_df['Auto_Cone_Mid'] + output_df['Tele_Cube_Mid'] + output_df['Tele_Cone_Mid'] + output_df['Auto_Cube_High'] + output_df['Auto_Cone_High'] + output_df['Tele_Cube_High'] + output_df['Tele_Cone_High'])
+    output_df['Assumed_Link_Points'] = 0
+
+    # Calculate mobility/station points
+    output_df['Mobility_Points'] = output_df['Mobility'].map({0: 0, 1: 3})
+    output_df['Auto_Station_Points'] = output_df['Auto_Station'].map({AutoStation.noPoints: 0, AutoStation.failed: 0, AutoStation.docked: 8, AutoStation.engaged: 12})
+    output_df['Tele_Station_Points'] = output_df['Tele_Station'].map({TeleStation.noPoints: 0, TeleStation.parked: 0, TeleStation.failed: 0, TeleStation.docked: 8, TeleStation.engaged: 12})
```

Comment:
Should `TeleStation.parked` be mapped to 2 points (as teams get 2 points for parking at the end of teleop?

Reply:
The problem is that failing to dock could be a park but it takes priority over parking so some parks will be scored and some parks won't be scored.
It would be simpler if parking were a separate toggle like mobility but I don't think we have time to do that.